src/assets/wise5/authoringTool/addNode/abstract-import-step/abstract-import-step.component.ts

```diff
@@ -0,0 +1,35 @@
+import { Directive } from '@angular/core';
+import { ActivatedRoute, Router } from '@angular/router';
+import { ConfigService } from '../../../services/configService';
+import { CopyNodesService } from '../../../services/copyNodesService';
+import { InsertNodesService } from '../../../services/insertNodesService';
+import { TeacherProjectService } from '../../../services/teacherProjectService';
+
+@Directive()
+export abstract class AbstractImportStepComponent {
+  constructor(
+    protected configService: ConfigService,
+    protected copyNodesService: CopyNodesService,
+    protected insertNodesService: InsertNodesService,
+    protected projectService: TeacherProjectService,
+    protected route: ActivatedRoute,
+    protected router: Router
+  ) {}
+
+  protected submit(
+    nodesToImport: any[],
+    importProjectId: number,
+    nodeIdToInsertInsideOrAfter: string
+  ): void {
+    this.copyNodesService
+      .copyNodes(nodesToImport, importProjectId, this.configService.getProjectId())
+      .subscribe((copiedNodes: any[]) => {
+        const nodesWithNewNodeIds = this.projectService.getNodesWithNewIds(copiedNodes);
+        this.insertNodesService.insertNodes(nodesWithNewNodeIds, nodeIdToInsertInsideOrAfter);
+        this.projectService.checkPotentialStartNodeIdChangeThenSaveProject().then(() => {
+          this.projectService.refreshProject();
+          this.router.navigate(['../../..'], { relativeTo: this.route });
+        });
+      });
+  }
```

Comment:
Rename function to import, and declare ```importProjectId``` and ```nodeIdToInsertInsideOrAfter``` as protected fields in this class? This way, you don't need to redeclare them in each of the child classes, and you won't need to pass them into this function.

```suggestion
  protected import(nodesToImport: any[]): void {
    this.copyNodesService
      .copyNodes(nodesToImport, this.importProjectId, this.configService.getProjectId())
      .subscribe((copiedNodes: any[]) => {
        const nodesWithNewNodeIds = this.projectService.getNodesWithNewIds(copiedNodes);
        this.insertNodesService.insertNodes(nodesWithNewNodeIds, this.nodeIdToInsertInsideOrAfter);
        this.projectService.checkPotentialStartNodeIdChangeThenSaveProject().then(() => {
          this.projectService.refreshProject();
          this.router.navigate(['../../..'], { relativeTo: this.route });
        });
      });
  }
```